### DISCUSSION TITLE
Sampling Decay Data

### DISCUSSION THREAD

**Discussion by EnricaBelfiore** [luca-fiorito-11/sandy#158] [General]
To complete the sampling of the decay data, in a consistent manner with respect to what is implemented for the cross sections, we have to include:

- methods for the perturbations of half life, branching ratios and decay energy;
- methods to create a covariance matrix or to add physical constrains (e.g. normalization of branching ratios) to the sampling population;
- method to extract best estimates and uncertainties of decay energies

**Comment by luca-fiorito-11** [comment]
@EnricaBelfiore , can you be more clear on which methods and how they should communicate in a pipeline?
@GrimFe @AitorBengoechea , any feedback is welcome...

**Reply by EnricaBelfiore** [reply]
Concerning the sampling procedure of decay data, I have developed a simple Notebook 
which shows the scheme of the algorithm for the perturbation of half life. This is how I'm thinking it for the moment: https://github.com/EnricaBelfiore/SandyProposals/blob/MyNotebook/Sampling_half_life.ipynb

As we can see in the notebook, the methods and functions that are needed for this purpose are:

- `get_endf6_file` to import the corresponding ENDF-6 nuclear data file directly from internet;
- `from_endf6` method of the DecayData class to extract the radioactive decay data information stored in 
the retrieved ENDF-6 file;
- `get_halflives` (or get_corresponding_decay_data, some of them are _not yet implemented_) to extract the best estrimates and their corresponding uncertainty. An option could be to include a *kwarg `relative` to obtain the relative or absolute value of the uncertainty;
- `get_covariance` which is _not yet implemented_. The idea is to develop a method that returns the covariance matrix of a decay data (as a `sandy.CategoryCov` object). This will be evaluated according with the uncertainty 
information (equal to the standard deviation) stored in ENDF6 and with the insertion of correlations calculated 
with deterministic method (or adding physical constains to the sample population  after the sampling);
- `sampling` method of CategoryCov or other option (the one that will be chosen as default method in SANDY
to extract perturbation coefficients with covariance from given covariance matrix);
- `custom_perturmation_HL` (or custom_perturbation_corresponding_decay_data), _which is not yet implemented_,
to apply the perturbation coefficients to the best estimates values and return updated DecayData objects with
the perturbed values;
- `to_endf6` method of the DecayData class to update the endf6 file with the perturbed information;
- `to_file` method of Endf6 class to eventually write the data into a file. 

This procedure will be adapted to be coherent with the updates for the other nuclear data pipeline.

**Reply by luca-fiorito-11** [reply]
Based on your directions I implemented a project board for [sampling decay data](https://github.com/luca-fiorito-11/sandy/projects/6). Please, adjust it as needed!
